File: ios/Runner/PyTorchYoloModel.swift
import LibTorch

class PyTorchYoloModel {
    var module: TorchModule?

    init(modelPath: String) {
        module = TorchModule(modelPath: modelPath)
    }

    func runInference(input: [Float]) -> DetectionResult {
        let inputTensor = Tensor(from: input, shape: [1, 3, 640, 640])
        let outputTensor = module?.forward(inputTensor) ?? Tensor()
        return parseOutput(outputTensor)
    }

    private func parseOutput(_ outputTensor: Tensor) -> DetectionResult {
        // Implement your parsing logic here
        return DetectionResult() // Replace with actual parsing logic
    }
}

struct DetectionResult {
    var boxes: [BoundingBox]
    var count: Int
}

struct BoundingBox {
    var x1, y1, x2, y2: Float
    var classId: Int
}

File: android/app/src/main/kotlin/com/example/fish_egg_realtime_detection/PyTorchYoloModel.java
import org.pytorch.IValue;
import org.pytorch.Module;
import org.pytorch.Tensor;

import java.util.ArrayList;
import java.util.List;

public class PyTorchYoloModel {
    private Module module;

    public PyTorchYoloModel(String modelPath) {
        module = Module.load(modelPath);
    }

    public DetectionResult runInference(float[] input) {
        Tensor inputTensor = Tensor.fromBlob(input, new long[]{1, 3, 640, 640});
        Tensor outputTensor = module.forward(IValue.from(inputTensor)).toTensor();
        return parseOutput(outputTensor);
    }

    private DetectionResult parseOutput(Tensor outputTensor) {
        // Implement your parsing logic here
        return new DetectionResult(); // Replace with actual parsing logic
    }
}

class DetectionResult {
    List<BoundingBox> boxes = new ArrayList<>();
    int count = 0;
}

class BoundingBox {
    float x1, y1, x2, y2;
    int classId;
}
